feat: TRR2 resolver

DecFox opened the pull request:
## Checklist

- [x] I have read the [contribution guidelines](https://github.com/ooni/probe-cli/blob/master/CONTRIBUTING.md)
- [x] reference issue for this pull request: https://github.com/ooni/probe/issues/2185
- [ ] if you changed anything related how experiments work and you need to reflect these changes in the ooni/spec repository, please link to the related ooni/spec pull request: <!-- add URL here -->

<!-- Reminder: Location of the issue tracker: https://github.com/ooni/probe -->

## Description

This intoduces a TRR2 resolver to measurexlite allowing us to opportunistcally test for DoH and make use of it whenever we can. The resolver uses the system resolver (`getaddrinfo`) as a fallback when DoH fails

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+)
+
+// NewTrustedRecursiveResolver2 returns a new trusted recursive resolver
```
```suggestion
// NewTrustedRecursiveResolver2 returns a new trusted recursive resolver.
```

bassosimone replied:
You should document that `logger` is `MANDATORY` while `URL` is `OPTIONAL` and you should say what happens if you provide an empty `URL`. You should probably say, for future robustness, that we're going to choose a suitable URL (or you could also be specific, up to you).

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+	}
+}
+
+// TrustedRecursiveResolver2 emulates Firefox's TRR2 mode
```
```suggestion
// TrustedRecursiveResolver2 emulates Firefox's TRR2 mode.
```

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+
+// TrustedRecursiveResolver2 emulates Firefox's TRR2 mode
+type TrustedRecursiveResolver2 struct {
+	Logger model.Logger
+	URL    string
```
Please, document `Logger` and `URL` (and mention they're `MANDATORY`)

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+// we can't return a decisive address here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Address() string {
+	return ""
```
How about returning the URL here?

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+// we can't return a decisive network here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Network() string {
+	return ""
```
How about returning `"trr2"` here?

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+}
+
+// CloseIdleConnections implements model.Resolver.CloseIdleConnections
+// we only check the fallback resolver since DoH resolver is generated on the fly
```
```suggestion
```

I think this is an implementation detail, so we can omit the comment

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+
+// LookupHost implements model.Resolver.LookupHost
+func (trr *TrustedRecursiveResolver2) LookupHost(ctx context.Context, hostname string) ([]string, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
```
I think this code should read like:

```Go
reso := trr.newParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
```

where `newParallelDNSOverHTTPSResolver` should read:

```Go
func (trr *TrustedRecursiveResolver2) newParallelDNSOverHTTPSResolver(logger model.Logger, URL string) model.Resolver {
  factory := trr.NewParallelDNSOverHTTPSResolverFn
  if factory == nil {
    factory = netxlite.NewParallelDNSOverHTTPSResolver
  }
  return factory(logger, URL)
}
```

DecFox replied:
You are right, this is more robust. Thanks!

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+
+// LookupHTTPS implements model.Resolver.LookupHTTPS
+func (trr *TrustedRecursiveResolver2) LookupHTTPS(ctx context.Context, domain string) (*model.HTTPSSvc, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
```
This code should be refactored as suggested for `LookupHost`

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+
+// LookupNS implements model.Resolver.LookupNS
+func (trr *TrustedRecursiveResolver2) LookupNS(ctx context.Context, domain string) ([]*net.NS, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
```
This code should be refactored as suggested for `LookupHost`

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+	return out, err
+}
+
+// Url returns the configured DoH URL for the TRR2 resolver
```
At this point while reading the diff, I am failing to understand why we need this method given that `URL` is already a public attribute of the structure and given that, if we use a `model.Resolver`, we cannot access the `Url` method anyway.

DecFox replied:
Sorry, this was actually some experimental code I forgot to remove. Thanks for pointing this out. I have removed it in the new commit.

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+
+//
+// Trusted Recursive Resolver
+//
```
Either call the file `trr2.go` and keep this comment or call this file `resolver.go` and say it contains resolvers

bassosimone commented on internal/measurexlite/resolver_test.go:
```diff
@@ -0,0 +1,163 @@
+)
+
+func TestNewTrustedRecursiveResolver(t *testing.T) {
+	var called bool
```
Please, put suitable copies of this variable in the narrowest possible scope to avoid mistakes caused by the variable in a subtest using a value previously computed by another subtest.

bassosimone replied:
(I see how in the code you only use it once, but I'd like to recommend you probably just create the resolver you need each time rather than having too much shared state for tests. The issue with shared state is that it's not robust to refactoring, so I'd rather just initialize each time the specific calls I need and leave others as `nil`)

bassosimone commented on internal/measurexlite/resolver_test.go:
```diff
@@ -0,0 +1,163 @@
+	})
+
+	t.Run("with DoH resolver", func(t *testing.T) {
+		resolver := NewTrustedRecursiveResolver2(&mocks.Logger{}, "")
+		resolvert := resolver.(*TrustedRecursiveResolver2)
```
```suggestion
		resolver := NewTrustedRecursiveResolver2(&mocks.Logger{}, "").(*TrustedRecursiveResolver2)
```

This should save you a variable.

DecFox replied:
You are right. Thanks!

bassosimone commented on internal/measurexlite/resolver_test.go:
```diff
@@ -0,0 +1,163 @@
+		resolvert := resolver.(*TrustedRecursiveResolver2)
+		resolvert.NewParallelDNSOverHTTPSResolverFn = newResolver
+		ctx := context.Background()
+		t.Run("LookupHost", func(t *testing.T) {
```
```suggestion

		t.Run("LookupHost", func(t *testing.T) {
```

Please always add spaces around `t.Run` invocations. This allows for easy navigation of blocks using vi's `{` and `}` commands.

bassosimone commented on internal/engine/probeservices/measurementmeta_test.go:
```diff
@@ -52,7 +52,7 @@ func TestGetMeasurementMetaWorkingAsIntended(t *testing.T) {
 		t.Fatal("unexpected input value")
 	}
-	if mmeta.MeasurementStartTime.String() != "2020-12-09 05:22:25 +0000 UTC" {
+	if mmeta.MeasurementStartTime.String() != "2020-12-09 04:22:25 +0000 UTC" {
```
I don't fully understand why this change is necessary and the commit does not seem to explain in detail why this was behaving in a flaky way. Would you mind clarifying?

DecFox replied:
There was a bug in the data pipeline (as I had informed in our 1:1). I have reverted this back since the bug has been fixed now: https://github.com/ooni/probe-cli/pull/846/commits/f71c4e29d5c413254ef6d37884abe2bf0a86bdfe

bassosimone reviewed (changes requested):
I like the work you did so far. As I mention to you in private, we probably need to figure out some design details around this functionality, which won't cause dramatic changes, so we can discuss them in parallel with the code review proper. I highlighted a bunch of places in which I think you should tweak/refactor your code for improved robustness.

Additionally, I've seen quite a large coverage drop (-0.5%) which most likely means that we have added a bit too much code that we're not testing. I would recommend for you to check which are the new functions that we're not testing and add tests.

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	addrs, err := reso.LookupHost(ctx, hostname)
```
I forgot to mention: if we want to implement TRR2 correctly, this lookup should be strictly time bounded (i.e., we should timeout the lookup after N seconds have elapsed)

DecFox replied:
I have added a timeout to the DoH resolver here: https://github.com/ooni/probe-cli/pull/846/commits/286c78f1b0a2fadabd2119f89180cea43d8c8476#diff-4c58fee97f07a0f47542cb08705a81d5f43f76be1d6235538ceb3e4c51725e72R33,
this is currently configured to use Firefox's default timeout value of 1.5s

bassosimone commented on internal/measurexlite/resolver.go:
```diff
@@ -0,0 +1,106 @@
+	}
+	out, err := reso.LookupHTTPS(ctx, domain)
+	if err != nil {
+		out, err = trr.ResolverSystem.LookupHTTPS(ctx, domain)
```
Because the system resolver only implements `LookupHost`, we should consider implementing `model.SimpleResolver` rather than `model.Resolver`.

DecFox replied:
> I like the work you did so far. As I mention to you in private, we probably need to figure out some design details around this functionality, which won't cause dramatic changes, so we can discuss them in parallel with the code review proper. I highlighted a bunch of places in which I think you should tweak/refactor your code for improved robustness.
> 
> Additionally, I've seen quite a large coverage drop (-0.5%) which most likely means that we have added a bit too much code that we're not testing. I would recommend for you to check which are the new functions that we're not testing and add tests.

Thank you for the feedback. I have made some changes based on the review. I also added some more tests but the coverage still seems to be decreased, so we probably need to add more tests.

bassosimone reviewed (changes requested):
Additional issues that I discovered while writing a design document for this feature

DecFox commented:
> I like the work you did so far. As I mention to you in private, we probably need to figure out some design details around this functionality, which won't cause dramatic changes, so we can discuss them in parallel with the code review proper. I highlighted a bunch of places in which I think you should tweak/refactor your code for improved robustness.
> 
> Additionally, I've seen quite a large coverage drop (-0.5%) which most likely means that we have added a bit too much code that we're not testing. I would recommend for you to check which are the new functions that we're not testing and add tests.

Thanks you for the feedback! I have made some changes based on the review. I have also added more tests, but the coverage still seems to be decreased. So we probably need to write some more.
